middleware/grpc_instrumentation_test.go

```diff
@@ -43,3 +55,260 @@ func TestErrorCode_Unknown(t *testing.T) {
 	a := errorCode(err)
 	assert.Equal(t, "error", a)
 }
+
+func setUpStreamClientInstrumentInterceptorTest(t *testing.T, serverStreams bool) (grpc.ClientStream, *mockGrpcStream, context.CancelFunc, *prometheus.Registry) {
+	reg := prometheus.NewPedanticRegistry()
+	metric := promauto.With(reg).NewHistogramVec(prometheus.HistogramOpts{
+		Name: "client_request_duration_seconds",
+	}, []string{"method", "result"})
+
+	interceptor := StreamClientInstrumentInterceptor(metric)
+	ctx, cancelCtx := context.WithCancel(context.Background())
+	t.Cleanup(cancelCtx)
+
+	mockStream := &mockGrpcStream{}
+	streamer := func(ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+		return mockStream, nil
+	}
+	desc := &grpc.StreamDesc{ServerStreams: serverStreams}
+	stream, err := interceptor(ctx, desc, nil, "/thing.Server/DoThing", streamer)
+	require.NoError(t, err)
+
+	return stream, mockStream, cancelCtx, reg
+}
+
+func TestStreamClientInstrumentInterceptor_SendMsg(t *testing.T) {
+	stream, mockStream, cancelCtx, reg := setUpStreamClientInstrumentInterceptorTest(t, true)
+
+	err := stream.SendMsg("message 1")
+	require.NoError(t, err)
+	requireNoRequestMetrics(t, reg)
+
+	mockStream.sendMsgError = errors.New("something went wrong")
+	err = stream.SendMsg("message 2")
+	require.Equal(t, mockStream.sendMsgError, err)
+	requireRequestMetrics(t, reg, "error", 1)
+
+	requireDoesNotIncrementMetricsAfterCompletion(t, stream, mockStream, cancelCtx, reg, "error")
+}
+
+func TestStreamClientInstrumentInterceptor_RecvMsg(t *testing.T) {
+	t.Run("server is streaming", func(t *testing.T) {
+		t.Run("RecvMsg returns an EOF", func(t *testing.T) {
+			stream, mockStream, cancelCtx, reg := setUpStreamClientInstrumentInterceptorTest(t, true)
+
+			err := stream.RecvMsg(nil)
+			require.NoError(t, err)
+			requireNoRequestMetrics(t, reg)
+
+			mockStream.recvMsgError = io.EOF
+			err = stream.RecvMsg(nil)
+			require.Equal(t, mockStream.recvMsgError, err)
+			requireRequestMetrics(t, reg, "2xx", 1)
+
+			requireDoesNotIncrementMetricsAfterCompletion(t, stream, mockStream, cancelCtx, reg, "2xx")
```

Comment:
This check that a 2xx is not tracked, but I think we should ensure no other status is tracked too.

_Comment applies below too._

Reply:
`gatherRequestMetrics` (which is used by `requireRequestMetrics`) returns all metrics, so this will catch metrics reported with the wrong status.